pkg/replica/hash.go

```diff
@@ -0,0 +1,220 @@
+		requireRehash = t.isRehashRequired(modTime)
+		if !requireRehash {
+			return
+		}
+	}
+
+	logrus.Infof("Starting hashing snapshot %v", t.SnapshotNames)
+
+	checksum, err = t.hashSnapshot()
+	if err != nil {
+		return
+	}
+
+	logrus.Infof("Snapshot %v checksum %v", t.SnapshotNames, checksum)
+
+	if err := t.setSnapshotHashInfoToXattr(checksum, modTime); err != nil {
+		return
+	}
+}
+
+func (t *SnapshotHashTask) openSnapshot() error {
+	t.Lock()
+	defer t.Unlock()
+
+	dir, err := os.Getwd()
+	if err != nil {
+		return errors.Wrapf(err, "cannot get working directory")
+	}
+
+	snapshotDiskNames := []string{}
+	for _, name := range t.SnapshotNames {
+		snapshotDiskNames = append(snapshotDiskNames, diskutil.GenerateSnapshotDiskName(name))
+	}
+
+	r, err := NewReadOnly(dir, snapshotDiskNames[0], snapshotDiskNames, nil)
+	if err != nil {
+		return err
+	}
+
+	t.replica = r
+
+	return nil
+}
+
+func (t *SnapshotHashTask) closeSnapshot() error {
+	t.Lock()
+	defer t.Unlock()
+
+	err := t.replica.Close()
+	t.replica = nil
+
+	return err
+}
+
+func (t *SnapshotHashTask) readSnapshot(start int64, data []byte) error {
+	_, err := t.replica.ReadAt(data, start)
+	return err
+}
+
+func (t *SnapshotHashTask) statSnapshot() (os.FileInfo, error) {
+	return os.Stat(diskutil.GenerateSnapshotDiskName(t.SnapshotNames[0]))
+}
+
+func (t *SnapshotHashTask) getSnapshotHashInfoFromXattr() (string, string, error) {
+	xattrSnapshotHashValue := make([]byte, xattrSnapshotHashValueMax)
+	_, err := unix.Getxattr(diskutil.GenerateSnapshotDiskName(t.SnapshotNames[0]), xattrSnapshotHashName, xattrSnapshotHashValue)
+	if err != nil {
+		return "", "", err
+	}
+
+	index := bytes.IndexByte(xattrSnapshotHashValue, 0)
+
+	info := &SnapshotXattrHashInfo{}
+	if err := json.Unmarshal(xattrSnapshotHashValue[:index], info); err != nil {
+		return "", "", err
+	}
+
+	return info.Checksum, info.ModTime, nil
+}
+
+func (t *SnapshotHashTask) setSnapshotHashInfoToXattr(checksum, modTime string) error {
+	xattrSnapshotHashValue, err := json.Marshal(&SnapshotXattrHashInfo{
+		Checksum: checksum,
+		ModTime:  modTime,
+	})
+	if err != nil {
+		return err
+	}
+
+	return unix.Setxattr(diskutil.GenerateSnapshotDiskName(t.SnapshotNames[0]), xattrSnapshotHashName, xattrSnapshotHashValue, 0)
+}
+
+func (t *SnapshotHashTask) getSize() int64 {
+	return t.replica.info.Size
+}
+
+func (t *SnapshotHashTask) isRehashRequired(currentModTime string) bool {
+	checksum, modTime, err := t.getSnapshotHashInfoFromXattr()
+	if err != nil {
+		logrus.Errorf("failed to get snapshot %v last hash info from xattr since %v", t.SnapshotNames[0], err)
+		return true
+	}
+
+	if modTime != currentModTime || checksum == "" {
+		return true
+	}
+
+	return false
+}
+
+func (t *SnapshotHashTask) hashSnapshot() (string, error) {
+	err := t.openSnapshot()
+	if err != nil {
+		logrus.Errorf("failed to open snapshot %v since %v", t.SnapshotNames, err)
+		return "", err
+	}
+	defer t.closeSnapshot()
+
+	h := sha256.New()
```

Comment:
Do we consider using CRC instead of sha256 in terms of performance perspective? 
https://stackoverflow.com/a/996873

Reply:
CRC is a non-cryptographic hash algorithm and cannot provide collision resistance guarantee. This mean it is unable to detect the file corruption or content change in some cases. We can also use a faster cryptographic algorithm such as MD5 for the tradeoff between performance impact and detection capability.

Actually, I'm fine with the two solutions.

@joshimoo @shuo-wu @innobead WDYT

Reply:
IIRC, zfs uses CRC for block error detection. I think we can use CRC in the implementation here.

Reply:
Updated, thank you.

Reply:
For knowledge sharing: CRC is a better approach for error detection, since it's less computationally expensive, but further it actually allows for detecting the individual bit/byte error based on the chosen CRC(size,polynominal) and the data size to be checked, it also can be done piece wise (different sections in parallel then merge).

It is correct that it's not secure against deliberate modifications and that the collision chance might be higher than with a cryptographic hash, but the chance for collisions can be increased/decreased by the choice of the CRC(size,polynominal) and the size of the data to be checked.

There is a 64bit package as well, but It doesn't look like the arm CPU has a dedicated instruction for it.
https://pkg.go.dev/hash/crc64@go1.19.2

Not sure of the top of my head on the intel side :)